### PR TITLE
Caching services and values for fast performance

### PR DESCRIPTION
**Pull request by Krellan**
Adding service cache, to avoid repeated calls to getService(), as the
returned information is unlikely to change. Any error resulting from
calling the service will clear this cache, in case the service name
actually did change.

Adding value cache, to cache the last good value received, from each
sensor. Using D-Bus message notifications, to become aware of changes
to a sensor, instead of polling the value every time through the main
loop. This avoids repeated polling for sensors which are maintaining
a stable temperature. Only good readings are cached, any invalid
reading also clears this cache.

An optional timeout feature is also added, in case message
notifications are not working well on your system. After the timeout,
the value cache for this sensor will be cleared, even if the reading
was good, so another poll will then be made.

A separate cache debugging flag is also added, so it can be enabled
independently of the main margin zone computation debugging flag.

Tested: Works nicely locally for me, sensor loop now is much faster.

Signed-off-by: Josh Lehan <krellan@google.com>

### CODE REVIEW

**Comment by Krellan** [comment]
Hi there, this patch should be ready to merge now. It is working nicely for me.

File: util/util.cpp [hunk 1]
```diff
@@ -106,41 +518,16 @@ double getSensorDbusTemp(std::string sensorDbusPath, bool unitMilli)
...
     if (spofEnabled)
     {
-        // If sensor functional property is false, set NaN to dbus and return NaN.
-        if (!dbus::SDBusPlus::checkFunctionalProperty(bus,
-                                                      service,
-                                                      sensorDbusPath))
-        {
-            return value;
-        }
+        std::string alarmText = checkSensorAlarmProperties(bus, service, sensorDbusPath);
```
Comment (harveyquta) [inline util/util.cpp]: Hi Krellan, sorry for the late review. Thanks to improve this daemon.
After testing, the behavior is correct in normal situation, but seems there is some problem when sensor fail.
When I trigger sensor warning/critical high, the value will set as NaN and fans will into fanfail duty in the beginning, but after few seconds, it will back to normal duty cycle.
If I understand correctly, does there need "setValueCache" when detecting alarmText not empty?

Comment (Krellan) [reply util/util.cpp]: Interesting, thanks for testing it. I don't use the warning/critical thresholds, so didn't test with those.
What commands do you use to trigger the warning/critical high?
Also, does it still work correctly without this patch in place? The intention of this patch is not to alter existing behavior, so if the behavior did change, then that's definitely a bug that we should fix.
The "setValueCache" function should not be called when there is an active alarm.
Reason is, the value returned should be NaN, so that the D-Bus reading will be done again during the next loop.
In other words, the caching behavior is defeated when there is an alarm, intentionally, as that is not a normal condition. By design, there is no "negative cache", so any alarm/error conditions will cause a fallback to the previous behavior of polling.

Comment (harveyquta) [reply util/util.cpp]: I use set-property to change sensor warning/critical high, and then will auto trigger the warning/critical thresholds.
Without this patch, target sensor value will always NaN when trigger the warning/critical thresholds until deassert the warning/critical thresholds, because the origin code will always check sensor thresholds and functional.
But including this patch, the target sensor value will into NaN just few seconds after trigger the warning/critical thresholds.
I think there is some problem in line472 (double value = getValueCache(sensorDbusPath, bus);).

Comment (harveyquta) [reply util/util.cpp]: Maybe the function "clearValueCache" have to use in some functions? I didn't see this function be used.

Comment (quadpixels) [reply util/util.cpp]: For the concern over warning/critical thresholds getting triggered, perhaps catching "CriticalHighAlarmAsserted" and "WarningHighAlarmAsserted" signals would help?

Comment (harveyquta) [reply util/util.cpp]: The signal warning/critical thresholds are already detected. But the problem is, after including this patch, the target dbus value will be NaN just a few seconds after trigger warning/critical thresholds, and then return to non-NaN. The origin code will back to non-NaN after detecting the sensor deasserted signal.

Comment (quadpixels) [reply util/util.cpp]: It seems on my test platform the sensors do not implement `xyz.openbmc_project.Sensor.Threshold.Critical` and `xyz.openbmc_project.Sensor.Threshold.Warning` interfaces and do not emit the `CriticalHighAlarmAsserted`, `CriticalLowAlarmAsserted`, `WarningHighAlarmAsserted` and `WarningLowAlarmAsserted` signals. I'm wondering what sensor daemons you are using?

Comment (harveyquta) [reply util/util.cpp]: I'm not sure what sensor daemon you are using.
If phosphor-hwmon, the sensor thresholds have to define in sensor's .conf files.
If entity-manager, it has to define in entity-manager's configuration files.

**Comment by Krellan** [comment]
Thanks for the feedback.

I have made a change, so that when an alarm condition is indicated, the value cache is invalidated. This will ensure the alarm condition is polled for each time through the loop, which will result in slower performance, but safer in the presence of an alarm condition: the alarm will ensure that the value remains NaN until the alarm condition is cleared. This better matches the behavior of the previous code before this patch.

Please test it again, and see if it works for you now.

**Comment by harveyquta** [comment]
Interesting thing.
After simple testing, when I ran read-margin-temp without any option and triggered threshold signal, the value didn't become NaN.
But when I added "--debug" option and triggered the critical threshold signal, the target value became NaN, and the behavior is like the previous code.
After checking this patch, I have no idea about this situation currently.

**Comment by Krellan** [comment]
Very strange. The debug option should not make a difference here. What commands are you using to test the thresholds with?
Can you reproduce the issue again, or did it happen only once?

File: util/util.cpp (hunk 1 above)
Comment (quadpixels) [reply util/util.cpp]: I see, I think the problem is in my setup, the phosphor-hwmon sensors are not configured with thresholds yet.

Do you see DBus signals that look like the following emitted from a phosphor-hwmon sensor, when a threshold is exceeded? (These can be seen with `dbus-monitor --system`)
```
signal time=1636003048.639621 sender=:1.47 -> destination=(null destination) serial=33
path=/xyz/openbmc_project/sensors/utilization/CPU;
interface=xyz.openbmc_project.Sensor.Threshold.Critical;
member=CriticalHighAlarmAsserted
   double 22.1945
```

It looks like dbus-sensors does not make use of those signals yet (because methods like `sdbusplus::xyz::openbmc_project::Sensor::Threshold::server::Warning::sdbusplus::xyz::openbmc_project::Sensor::Threshold::server::Warning` do not appear in dbus-sensors). It does appear in phosphor-hwmon though.

So I guess for this optimization to work with dbus-sensors correctly, dbus-sensors needs to be updated to emit those signals when thresholds are crossed.

**Comment by harveyquta** [comment]
My bad. After detailed testing, it seems still have some problems when triggering the threshold.
Here is the detailed testing log, [Normal Mode](https://drive.google.com/file/d/1hcZUUBiaEzS4lTixMb8lox8HQObFL9yN/view?usp=sharing), [Debug Mode](https://drive.google.com/file/d/1SbuOxENe6F-iTZQbasdwl8RnpdBzdj7Q/view?usp=sharing)
The NaN value will only exist for about 10 seconds.

**Comment by harveyquta** [comment]
This situation seems not happen each times. Sometimes happen, sometimes didn't happen

**Comment by harveyquta** [comment]
Another testing again, NaN was keeping about 45 seconds and then back to 5X value. It's really strange....

**Comment by harveyquta** [comment]
After debuging with dbus-monitor, I think there is some problem in function "parseMessage".
When trigger threshold, this function will return NaN
But when trigger value propertyChanged next time, this function will update the value reading, so that the target value will become not NaN.

**Comment by Krellan** [comment]
Thanks for testing it. I think I have identified the problem: changes to the value are monitored by polling and by signal interrupts, however, changes to the alarm flags are monitored by polling only. Not by signal interrupts. So, they can get out of sync easily. Also, the state of the alarm flag is not maintained, another cause to get out of sync. I will take a look at the design and make it more correct with respect to handling alarm flags.

**Comment by Krellan** [comment]
I added a new feature. Now, alarms are handled better. Incoming alarm notifications are cached, just like incoming values. Whenever an alarm indicates a bad condition, the value will be held at NaN, and polling will be used to determine when all the alarms are cleared. When the alarm is cleared, and conditions are back to normal good condition, the value will once again be correct from the sensor. Hopefully this should restore correct behavior.

**Comment by Krellan** [comment]
Please test this again. Local testing gives me confidence that this will work now. When a sensor is in alarm status, the value is overridden and becomes NaN (to force fans into failsafe mode). When the alarm ceases, the value is restored.

**Comment by harveyquta** [comment]
Thanks for improving. After testing, the alarm behavior is correct. I think this PR is already to merge.

File: util/util.cpp (hunk 1 above)
Comment (quadpixels) [reply util/util.cpp]: Update: it seems we don't need the `Asserted` signals because the same assertion information can be checked in the `PropertyChanged` signal handlers. We also checked a few dbus dumps with and without the patch applied and we can see the program functioning as intended and the number of GetAll method calls greatly reduced with the improvement.
